can/bus.py

```diff
@@ -37,29 +51,58 @@ def __init__(self, channel=None, can_filters=None, **config):
             The can interface identifier. Expected type is backend dependent.
 
         :param list can_filters:
-            A list of dictionaries each containing a "can_id", a "can_mask",
-            and an "extended" key.
-
-            >>> [{"can_id": 0x11, "can_mask": 0x21, "extended": False}]
-
-            A filter matches, when ``<received_can_id> & can_mask == can_id & can_mask``
+            See :meth:`~can.BusABC.set_filters` for details.
 
         :param dict config:
             Any backend dependent configurations are passed in this dictionary
         """
-        pass
+        self.set_filters(can_filters)
 
     def __str__(self):
         return self.channel_info
 
-    @abstractmethod
     def recv(self, timeout=None):
         """Block waiting for a message from the Bus.
 
         :param float timeout: Seconds to wait for a message.
 
         :return:
             None on timeout or a :class:`can.Message` object.
+        :raise: :class:`can.CanError`
+            if an error occurred while reading
+        """
+        start = time()
+
+        while True:
+
+            # try to get a message
+            msg, already_filtered = self._recv_internal(timeout=timeout)
+
+            # return it, if it matches
+            if msg and (already_filtered or self._matches_filters(msg)):
+                return msg
+
+            # if not, and timeout is None, try indefinitely
+            elif timeout is None:
+                continue
+
+            # try next one only if there still is time, and with reduced timeout
+            else:
+
+                timeout = timeout - (time() - start)
+
+                if timeout > 0:
+                    continue
+                else:
+                    return None
+
+    @abstractmethod
```

Comment:
Don't make this abstract or it will brake older external implementations. If an interface implements `recv()` directly then this will also work.

Reply:
You are right, I changed it.